lib/rspec/matchers/expecteds_for_multiple_diffs.rb

```diff
@@ -0,0 +1,118 @@
+      end
+
+      # @api private
+      # Returns message with diff(s) appended for provided differ
+      # factory and actual value if there are any
+      # @param [String] message original failure message
+      # @param [Proc] differ_factory brand new differ factory method (or proc, block)
+      # @param [Any] actual value
+      # @return [String]
+      def message_with_diff(message, differ_factory, actual)
+        diff = diffs(differ_factory, actual)
+        message = "#{message}\n#{diff}" unless diff.empty?
+        message
+      end
+
+    protected
+
+      # @api private
+      # Returns list of expected values.
+      # @return [Array<RSpec::Matchers::ExpectedsForMultipleDiffs::Expected>]
+      def expected_list
+        @_expected_list ||= []
+      end
+
+    private
+
+      def diffs(differ_factory, actual)
+        expected_list.map do |expected|
+          expected.diffs(differ_factory, actual)
+        end.inject([], :+).join("\n")
+      end
+
+      def expected_from(value, for_matcher)
+        return value.expected_list if self.class === value
+        [Expected.new(value, for_matcher)]
+      end
+
+      Expected = Struct.new(:value, :for_matcher)
+      # @api private
+      # Represents expected value with possible knowledge of original
+      # matcher
+      class Expected
+        # @api private
+        # Returns diffs for provided differ factory and actual value
+        # @param [Proc] differ_factory brand new differ factory method (or proc, block)
+        # @param [Any] actual value
+        # @return [String]
+        def diffs(differ_factory, actual)
+          diff = differ_factory.call.diff(actual, value)
+          return [] if diff.empty?
+          ["#{for_matcher.diff_headline}#{diff}"]
+        end
+      end
+
+      ForMatcher = Struct.new(:matcher)
+      # @api private
+      # Represents a matcher diff is being built for
+      class ForMatcher
+        # @api private
+        # Wraps provided matcher in ForMatcher. If it is already a
+        # ForMatcher just returns it.
+        # @param [RSpec::Matchers::BuiltIn::BaseMatcher, RSpec::Matchers::ExpectedsForMultipleDiffs::ForMatcher] matcher or for_matcher value
+        # @return [RSpec::Matchers::ExpectedsForMultipleDiffs::ForMatcher]
+        def self.from(matcher)
+          return matcher if self === matcher
+          new(matcher)
+        end
+
+        # @api private
+        # Returns proper diff headline for wrapper matcher with
+        # matcher description included
+        # @return [String]
+        def diff_headline
+          "Diff for (#{matcher.description}):"
+        end
+      end
+
+      # @api private
+      # Null object for ForMatcher, used in case for_matcher was not
+      # provided
+      class NoMatcher < ForMatcher
+        # @api private
+        # Returns default diff headline with no matcher description
+        # @return [String]
+        def diff_headline
+          "Diff:"
+        end
+      end
```

Comment:
`NoMatcher` doesn't really need to inherit from `ForMatcher`, right?  Seems like unnecessary inheritance.  In fact, `NoMatcher` is stateless, so IMO we'd be better off with:

``` ruby
module NoMatcher
  def self.diff_headline
    "Diff:"
  end
end
```

...and then changing `NoMatcher.new` to `NoMatcher` above.  As a singleton, stateless object, we'll have fewer object allocations which translates into less GC, etc.